### PR TITLE
Mx 15964 bridge service update

### PR DESCRIPTION
**Pull request by axenteoctavian**
Update to use nonce handler v3

### CODE REVIEW

**Review by sasurobert — DISMISSED** [review]

File: server/cmd/server/.env [hunk 1]
```diff
@@ -8,12 +8,12 @@ WALLET_PATH="wallet.pem"
...
 MULTIVERSX_PROXY="https://testnet-gateway.multiversx.com"
-# Multi Sig address on MultiversX to register the transactions
-MULTI_SIG_SC_ADDRESS="erd1spyavw0956vq68xj8y4tenjpq2wd5a9p2c6j8gsz7ztyrnpxrruqzu66jx"
+# Header verifier address on MultiversX to register the transactions
+HEADER_VERIFIER_SC_ADDRESS="erd1spyavw0956vq68xj8y4tenjpq2wd5a9p2c6j8gsz7ztyrnpxrruqzu66jx"
 # ESDT Safe address on MultiversX to execute the transactions
 ESDT_SAFE_SC_ADDRESS="erd1spyavw0956vq68xj8y4tenjpq2wd5a9p2c6j8gsz7ztyrnpxrruqzu66jx"
-# Max retries to wait in seconds for account nonce update after sending bridge txs
-MAX_RETRIES_SECONDS_WAIT_NONCE=60
+# Interval in milliseconds between sending bridge txs
```
Comment (mariusmihaic) [inline server/cmd/server/.env:15]: Miliseconds or seconds?

File: server/txSender/dataFormatter.go [hunk 2]
```diff
@@ -38,7 +38,7 @@ func (df *dataFormatter) CreateTxsData(data *sovereign.BridgeOperations) [][]byt
 	}
 
 	for _, bridgeData := range data.Data {
-		log.Debug("creating tx data", "bridge op hash", bridgeData.Hash)
+		log.Debug("creating tx data", "bridge op hash", bridgeData.Hash, "nr of operations", len(bridgeData.OutGoingOperations))
```
Comment (mariusmihaic) [inline server/txSender/dataFormatter.go:41]:
```suggestion
		log.Debug("creating tx data", "bridge op hash", bridgeData.Hash, "no. of operations", len(bridgeData.OutGoingOperations))
```

File: server/txSender/factory.go [hunk 3]
```diff
@@ -28,9 +28,9 @@ func CreateTxSender(wallet core.CryptoComponentsHolder, cfg TxSenderConfig) (*tx
 		return nil, err
 	}
 
-	nonceHandler, err := nonceHandlerV2.NewNonceTransactionHandlerV2(nonceHandlerV2.ArgsNonceTransactionsHandlerV2{
-		Proxy:            proxy,
-		IntervalToResend: time.Second * time.Duration(cfg.MaxRetriesSecondsWaitNonce),
+	nonceHandler, err := nonceHandlerV3.NewNonceTransactionHandlerV3(nonceHandlerV3.ArgsNonceTransactionsHandlerV3{
+		Proxy:          proxy,
+		IntervalToSend: time.Millisecond * time.Duration(cfg.IntervalToSend),
```
Comment (mariusmihaic) [inline server/txSender/factory.go:33]: Is it ok to use 1 milisecond?

File: server/txSender/txSender.go [hunk 4]
```diff
@@ -71,7 +71,7 @@ func checkArgs(args TxSenderArgs) error {
 	if check.IfNil(args.TxNonceHandler) {
 		return errNilNonceHandler
 	}
-	if len(args.SCMultiSigAddress) == 0 {
+	if len(args.SCHeaderVerifierAddress) == 0 {
 		return errNoMultiSigSCAddress
```
Comment (mariusmihaic) [inline server/txSender/txSender.go:75]: This returned error should be refactored as well

File: server/txSender/txSender.go [hunk 5]
```diff
@@ -122,9 +122,10 @@ func (ts *txSender) createAndSendTxs(ctx context.Context, data *sovereign.Bridge
 			}
 		default:
 			log.Error("invalid tx data received", "data", string(txData))
+			continue
```
Comment (mariusmihaic) [inline server/txSender/txSender.go:125]: I would say that this is a fatal error and should not continue...
Not very sure about it, though...

File: server/txSender/txSender_test.go [hunk 6]
```diff
@@ -127,7 +127,8 @@ func TestTxSender_SendTxs(t *testing.T) {
 		},
 	}
 	args.TxNonceHandler = &testscommon.TxNonceSenderHandlerMock{
-		ApplyNonceAndGasPriceCalled: func(ctx context.Context, address core.AddressHandler, tx *transaction.FrontendTransaction) error {
+		ApplyNonceAndGasPriceCalled: func(ctx context.Context, txs ...*transaction.FrontendTransaction) error {
+			require.Len(t, txs, 1) // we update transactions one at a time
```
Comment (mariusmihaic) [inline server/txSender/txSender_test.go:131]: Maybe...
Should we refactor to create all txs and then send them all in bulk?

File: server/cmd/server/.env (hunk 1 above)
Comment (axenteoctavian) [reply server/cmd/server/.env:15]: This the interval between the txs if we would send them in bulk. We don't want to wait seconds between transfers

File: server/txSender/factory.go (hunk 3 above)
Comment (axenteoctavian) [reply server/txSender/factory.go:33]: Yes, because in `.env` file the interval is in milliseconds

File: server/txSender/txSender_test.go (hunk 6 above)
Comment (axenteoctavian) [reply server/txSender/txSender_test.go:131]: I think I tried this and didn't work as expected, not sure if it was my bad or the sdk didn't work properly when spammed.
I'd say we keep it one at the time for now.

File: server/txSender/txSender.go (hunk 5 above)
Comment (mariusmihaic) [reply server/txSender/txSender.go:125]: It is better, but I meant we should directly return error, not continue.
This is fatal, we should not skip and send "half txs"

File: server/txSender/txSender_test.go (hunk 6 above)
Comment (mariusmihaic) [reply server/txSender/txSender_test.go:131]: Can we add a task to check if we send them in bulk?

It's better logic anyway... have them all formatted + send them all;
Instead, right now, we can format 2/3 correct + send them, but fail for the third one...

I think it's better to use the ""all or nothing" approach.

**Review by mariusmihaic — APPROVED** [review]
:green_circle: for now, we should refactor to send txs in bulk, w.r.t to open comments

**Review by sasurobert — APPROVED** [review]

**Pull request merged**